File: Tree/513. Find Bottom Left Tree Value.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var findBottomLeftValue = function(root) {    // Bottom left node means we need to traverse the given tree with BSF way and get the first one in the bottom layer
    const stack = [root];
    let value;
    let current = 0;
    let count = 1;
    while(stack.length) {
        const node = stack.shift();
        if (node) {
            if (current === 0) {
                value = node.val;
            }
            current++;
            if (node.left) {
                stack.push(node.left);
            }
            if (node.right) {
                stack.push(node.right);
            }
            if (current === count) {
                current = 0;
                count = stack.length;
            }
        }
    }
    return value;
};

File: Golang/Tree/513.Find-bottom-left-tree-value.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findBottomLeftValue(root *TreeNode) int {
	stack := []*TreeNode{root}
	for len(stack) > 0 {
		nextLevel := []*TreeNode{}
		for i := 0; i < len(stack); i++ {
			node := stack[i]
			if node.Left != nil {
				nextLevel = append(nextLevel, node.Left)
			}
			if node.Right != nil {
				nextLevel = append(nextLevel, node.Right)
			}
		}
		if len(nextLevel) == 0 {
			return stack[0].Val
		}
		stack = nextLevel
	}
	return -1
}
